### PR TITLE
Zachmu/turbine perf

### PR DESCRIPTION
**Pull request by zachmu**
1) Do pushdown analysis within subqueries
2) Push index lookups down to tables in subqueries

### CODE REVIEW

**Comment by zachmu** [comment]
This needs a bit more testing, but you can take a look now.

**Comment by zachmu** [comment]
Aaron, PTAL when you have a minute. Ended up expanding this quite a bit from the original intent to fix bugs it uncovered. The big detour was discovering the bad behavior in the analyzer regarding plans that never converge on successive analyzer passes. Used to be ignored, is now an error. Once I fixed that, found several functions that aren't comparable with reflect.deepEquals and were therefore always giving us 1000 analyzer passes, so fixed those.

File: sql/analyzer/optimize_joins.go [hunk 1]
```diff
@@ -127,7 +128,8 @@ func transformJoins(
 			// TODO: should we just do this for every query plan as a final part of the analysis?
 			//  This would involve enforcing that every type of Node implement Expressioner.
 			a.Log("transforming node of type: %T", node)
-			return FixFieldIndexesForExpressions(node)
+			// TODO: need the scope here
```
Comment (reltuk) [inline sql/analyzer/optimize_joins.go:131]: Current dangers?

File: sql/analyzer/pushdown.go [hunk 2]
```diff
@@ -16,81 +16,49 @@ func pushdownFilters(ctx *sql.Context, a *Analyzer, n sql.Node, scope *Scope) (s
...
+	if !canProject(n, a) {
 		return n, nil
 	}
 
-	return transformPushdownProjections(ctx, a, n)
+	return transformPushdownProjections(ctx, a, n, scope)
 }
 
-func canProject(n sql.Node) bool {
+func canProject(n sql.Node, a *Analyzer) bool {
```
Comment (reltuk) [inline sql/analyzer/pushdown.go:57]: Just curious why we threaded this here and in `canDoPushdown`...

**Review by reltuk — APPROVED** [review]
Generally looks good!

File: sql/analyzer/pushdown.go (hunk 2 above)
Comment (zachmu) [reply sql/analyzer/pushdown.go:57]: For logging. But it was unused in one of those places, removed.

File: sql/analyzer/optimize_joins.go (hunk 1 above)
Comment (zachmu) [reply sql/analyzer/optimize_joins.go:131]: Argh

File: enginetest/memory_engine_test.go [hunk 3]
```diff
@@ -73,20 +73,19 @@ func TestQueriesSimple(t *testing.T) {
 
 // Convenience test for debugging a single query. Unskip and set to the desired query.
```
[the remark this answers is not in the document]
Comment (zachmu) [reply enginetest/memory_engine_test.go]: Fixed

**Pull request merged**